code/modules/power/lighting.dm

```diff
@@ -394,25 +406,53 @@
 	var/area/A = get_area(src)
 	return A.lightswitch && A.power_light
 
-/obj/machinery/light/proc/flicker(amount = rand(10, 20))
-	if(flickering)
+///flicker lights on and off
+/obj/machinery/light/proc/flicker(toggle_flicker = FALSE)
+	if(!has_power())
+		addtimer(CALLBACK(src, PROC_REF(flicker)), flicker_time)
 		return
-	flickering = TRUE
-	spawn(0)
-		if(light_on && status == LIGHT_OK)
-			for(var/i = 0; i < amount; i++)
-				if(status != LIGHT_OK)
-					break
-				update(FALSE)
-				sleep(rand(5, 15))
-			update(FALSE)
+	if(toggle_flicker)
+		if(status != LIGHT_OK)
+			return
+		flickering = flickering? FALSE : TRUE
+		if(flickering)
+			lightambient = new(null, FALSE)
```

Comment:
this is never deleted, so you're just creating a new one every time. Again, if this is a component you'd just create it once when connected to a light, then you just start and stop it as needed.